[FLINK-21883][scheduler] Implement cooldown period for adaptive scheduler

echauchot opened the pull request:
## What is the purpose of the change

Introduce the cooldown period in the adaptive scheduler ([FLIP-322](https://cwiki.apache.org/confluence/display/FLINK/FLIP-322+Cooldown+period+for+adaptive+scheduler))

## Brief change log
Change Executing state, when new slots arrive:

-  schedule a rescale when last rescale was done less than interval-min seconds ago, rescale immediately otherwise.
-  Add the ability to force a rescale when last rescale was done more than interval-max seconds ago regardless if the requirements are met (context#shouldRescale)


## Verifying this change

Added 3 tests to ExecutingTest test case.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: adding to parameters related to timing
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: yes: adaptive scheduler in JobManager
  - The S3 file system connector:no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? javadoc

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -488,6 +488,23 @@ public enum SchedulerType {
                                             code(SchedulerExecutionMode.REACTIVE.name()))
                                     .build());
 
+    @Documentation.Section(Documentation.Sections.EXPERT_SCHEDULING)
+    public static final ConfigOption<Long> SCHEDULER_SCALING_INTERVAL_MIN =
+            key("jobmanager.adaptive-scheduler.scaling-interval.min")
+                    .longType()
+                    .defaultValue(30L) // favor lower scaling-interval.min for more reactive
+                    // rescaling and let the user increase the value for high workloads
+                    .withDescription(
+                            "Determines the minimum time (in seconds) between scaling operations in reactive mode.");
+
+    @Documentation.Section(Documentation.Sections.EXPERT_SCHEDULING)
+    public static final ConfigOption<Long> SCHEDULER_SCALING_INTERVAL_MAX =
+            key("jobmanager.adaptive-scheduler.scaling-interval.max")
+                    .longType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Allow the user to configure the time (in seconds) after which a scaling operation is triggered regardless if the requirements are met");
```
These 2 option types are better to update from long to Duration.

echauchot replied:
:+1:

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -39,14 +42,22 @@
 
     private final Context context;
+    private Instant lastRescale = Instant.EPOCH;
+    private final ScheduledExecutorService scheduledExecutor = Executors.newScheduledThreadPool(1);
```
Should we create a new thread pool or use a old thread pool?

If we must create a new thread pool, it's better to add a thread factory.

echauchot replied:
:+1: yes I could use Context.componentMainThreadExecutor

echauchot replied:
I also changed the tests: I no more schedule (no time based test to avoid race conditions) but rather set/check the rescale status

echauchot commented:
@1996fanrui thanks for your review. PTAL

echauchot commented:
Hi @1996fanrui any updates ?

echauchot commented:
Hi @1996fanrui any updates ?

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -144,6 +188,17 @@ private void maybeRescale() {
+    }
+
+    private void rescaleWhenCooldownPeriodIsOver() {
+        context.runIfConditionOrSchedule(
+                timeSinceLastRescale().compareTo(scalingIntervalMin) > 0,
+                this::maybeRescale,
+                scalingIntervalMin); // reset cooldown period
```
Schedule the `maybeRescale` after  `scalingIntervalMin` might not make sense.

For example, `scalingIntervalMin` is 30 seconds, `lastRescale` is `08:01:05`, and now is `08:01:10`.

From your code, the `maybeRescale` is called at `08:01:40`, however it should be called at `08:01:35`, right?

echauchot replied:
That is right. My first implementation scheduled ad 08:01:35 but @dmvk suggested [here](https://lists.apache.org/thread/m2w2xzfjpxlw63j0k7tfxfgs0rshhwwr) that we keep the min interval low but we restart the cooldown period each time. That is what is expressed in the FLIP

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -210,6 +265,15 @@ interface Context
+         * @param action action to run
+         * @param delay delay after which to run the action if the condition is not met
+         */
+        void runIfConditionOrSchedule(boolean condition, Runnable action, Duration delay);
```
This interface isn't necessary, if `timeSinceLastRescale().compareTo(scalingIntervalMin)` is true, we can call  `maybeRescale` directly.

If it's false, it's better to call `runIfState` to schedule. 

BTW, if the state isn't this(Executing), the `maybeRescale` shouldn't be called. That's why the `runIfState` name includes `IfState`. The `runIfConditionOrSchedule`  doesn't consider the `IfState`, it might have bug.

echauchot replied:
Agree, thanks

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -144,6 +188,17 @@ private void maybeRescale() {
+        return Duration.between(lastRescale, Instant.now());
+    }
+
+    private void rescaleWhenCooldownPeriodIsOver() {
```
`rescaleWhenCooldownPeriodIsOver` should consider one case that it's called in multiple times.

For example, scalingIntervalMin is 30 seconds, `lastRescale` is 08:01:05.

- The `rescaleWhenCooldownPeriodIsOver` is called at  08:01:10, it will schedule `maybeRescale` at `08:01:35`.
- The `rescaleWhenCooldownPeriodIsOver` maybe called at  08:01:15 again, it will schedule `maybeRescale` at `08:01:35` again.

The second or repeated schedule should be ignored.

echauchot replied:
ok thanks for pointing out.

1996fanrui replied:
If we don't drop it or don't have any filter, the `scaling-interval.min` cannot be guaranteed, right?

The interval of  `08:01:40` and `08:01:45` is 5 s, and the `scaling-interval.min`  is 30s. It may not be as expected. Do you think so?

echauchot replied:
Exactly. This is what I meant and implemented in last commit of yesterday : the call at 08:01:10 will schedule `maybeRescale` at 08:01:40 and the call at 08:01:15 will not entail any rescale. When `maybeRescale` is executed at  08:01:40, it will use the resources added at 08:01:15 when calling `Context#getExecutionGraph()`. So that seems fine.

1996fanrui reviewed (changes requested):
Hi @echauchot , sorry for the late review. I just reviewed the producation code, and doesn't review the test code so far. 

I left some comments, please take a look in your free time. And please let me know if any comment is wrong, thanks~

echauchot commented:
@1996fanrui I addressed all  your comments and I think we are in sync. Can you resume you review ?

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -55,7 +63,9 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
-            List<ExceptionHistoryEntry> failureCollection) {
+            List<ExceptionHistoryEntry> failureCollection,
+            Duration scalingIntervalMin,
+            Duration scalingIntervalMax) {
```
```suggestion
            @Nullable Duration scalingIntervalMax) {
```

echauchot replied:
+1 thanks for pointing out.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -124,17 +154,33 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
     private void maybeRescale() {
-        if (context.shouldRescale(getExecutionGraph())) {
-            getLogger().info("Can change the parallelism of job. Restarting job.");
+        final Duration timeSinceLastRescale = timeSinceLastRescale();
+        rescaleScheduled = false;
+        final boolean shouldForceRescale =
+                (scalingIntervalMax != null)
+                        && (timeSinceLastRescale.compareTo(scalingIntervalMax) > 0)
+                        && (lastRescale != Instant.EPOCH); // initial rescale is not forced
+        if (shouldForceRescale || context.shouldRescale(getExecutionGraph())) {
+            if (shouldForceRescale) {
+                getLogger()
+                        .info(
+                                "Time since last rescale ({}) >  {} ({}). Force-changing the parallelism of the job. Restarting the job.",
+                                timeSinceLastRescale,
+                                JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                                scalingIntervalMax);
+            } else {
+                getLogger().info("Can change the parallelism of the job. Restarting the job.");
+            }
+            lastRescale = Instant.now();
             context.goToRestarting(
                     getExecutionGraph(),
```
The implementation of  `scalingIntervalMax` may be wrong.

As I understand, `scalingIntervalMax` want to trigger a rescale when the `jobmanager.adaptive-scheduler.min-parallelism-increase` isn't met, right?

If yes, we should set a timer to call the `rescale()` after `scalingIntervalMax`, right? If we don't set a timer, the `rescale` won't be called after `scalingIntervalMax`, and no one to force rescale.

Also, we should add a test for this case.

echauchot replied:
in the FLIP we said that the triggering event of the rescale is resource addition (`Executing#onNewResourcesAvailable`) (downscale is triggered by failures) so there is no need to schedule: when such an event happens `maybeRescale` is always called (either directly or scheduled) and when it is called, the rescale is done if `timeSinceLastRescale > scalingIntervalMax` no matter the min resources.  

There is also a test: `ExecutingTest#testNotifyNewResourcesAvailableWithNoResourcesAndForce`

1996fanrui replied:
Thanks for the clarification!

Do you mean we force trigger a rescale when new resources comes and the `current time > lastRescaleTime + scalingIntervalMax`?

If so, I think the logic related to `scalingIntervalMax` should be moved into `maybeRescale`. `onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver`, when user changed the parallelism, `maybeRescale` will force trigger a rescale as well. right?

echauchot replied:
> Do you mean we force trigger a rescale when new resources comes and the current time > lastRescaleTime + scalingIntervalMax?

yes, that is what is what the community agreed on in the FLIP.

> If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.

This logic is already in maybeRescale() method

> onNewResourceRequirements also call the rescaleWhenCooldownPeriodIsOver, when user changed the parallelism; maybeRescale will force trigger a rescale as well. right?

yes onNewResourceRequirements is ultimately called by a rest endpoint to react to use action on the web ui so it will call maybeRescale in the end but that would not always mean a rescale: if timeSinceLastRescale < scalingIntervalMax and minimal resources are not met then there will be no rescale.

1996fanrui replied:
> > If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.
> 
> This logic is already in maybeRescale() method

Sorry, I have a typo here. I mean the logic related to scalingIntervalMax should be moved out of the maybeRescale. 

`onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver` when user changed the parallelism, maybeRescale shouldn't force trigger a rescale in this case even if the `current time > lastRescaleTime + scalingIntervalMax`.

Because after requirement is adjusted, resources cannot be ready immediately. The rescale is meaningless at this time even if the `current time > lastRescaleTime + scalingIntervalMax`.

echauchot replied:
> > > If so, I think the logic related to scalingIntervalMax should be moved into maybeRescale.
> > 
> > 
> > This logic is already in maybeRescale() method
> 
> Sorry, I have a typo here. I mean the logic related to scalingIntervalMax should be moved out of the maybeRescale.
> 
> `onNewResourceRequirements` also call the `rescaleWhenCooldownPeriodIsOver` when user changed the parallelism, maybeRescale shouldn't force trigger a rescale in this case even if the `current time > lastRescaleTime + scalingIntervalMax`.
> 
> Because after requirement is adjusted, resources cannot be ready immediately. The rescale is meaningless at this time even if the `current time > lastRescaleTime + scalingIntervalMax`.

Agree, I think the same is true for maybeRescale() called in the constructor to check if new resources have arrived during the starting process

echauchot replied:
> > > Do you mean we force trigger a rescale when new resources comes and the current time > lastRescaleTime + scalingIntervalMax?
> > 
> > 
> > yes, that is what is what the community agreed on in the FLIP.
> 
> I think this strategy just solve one case: the last resource comes after scalingIntervalMax. The job can rescale directly.
> 

It was made after this observation, the longer the pipeline runs, the more the resource gain is worth the restarting time. 

> However, it cannot solve the case: the last resource comes before scalingIntervalMax. Assuming the scalingIntervalMax is 5 miuntes, the `jobmanager.adaptive-scheduler.min-parallelism-increase` is 2, and the expected paralleslim is 100.
> 
> * The job has 99 TMs at 09:00:00, it run with parallelism 99.
> * At 09:03:00, the last TM comes.
> * The `Executing` will ignore the rescale due to the parallelism diff is 1, it less than `min-parallelism-increase` and `current time < lastRescaleTime + scalingIntervalMax`.
> 
> And then the job doesn't need the new resource and the expected parallelism isn't change, so the job won't rescale anymore.

It is not linked to the scalingIntervalMax, before this PR in such a case there was no rescale either because simply the minimum resource addition was not met.  It is by design of the min-parallelism-increase that we don't change

1996fanrui replied:
> It is by design of the min-parallelism-increase that we don't change

This design was changed for case1, right? 

If yes, the `scalingIntervalMax` wants to ignore the `min-parallelism-increase` if one job doesn't rescale for a long time. I think the semantic of scalingIntervalMax should be fixed to the user side and should not be related to the calling order in the code or the order in which resources arrive.

echauchot replied:
> I think the semantic of scalingIntervalMax should be fixed to the user side and should not be related to the calling order in the code or the order in which resources arrive.

I don't understand what you mean by "fixed to the user side".

echauchot replied:
Here is the summary of the offline discussion @1996fanrui and I had about `scalingIntervalMax` and forcing rescale (current code: when `added resources < min-parallelism-increase` we force a rescale if `timeSinceLastRescale > scalingIntervalMax`):
1. aim: the longer the pipeline runs, the more the (small) resource gain is worth the restarting time.
2. corner case: a resource `< min-parallelism-increase` arrives when `timeSinceLastRescale < scalingIntervalMax` and the pipeline is running for a long time (typical case 1) => with the current code, we don't force a rescale in that case whereas the added resource would be worth the restarting time.
=> I proposed solution 1: changing the definition of  `scalingIntervalMax` to `pipelineRuntimeRescaleMin` meaning pipeline runtime after which we force a rescale even if `added resources < min-parallelism-increase`
=> @1996fanrui  proposed solution 2: if `added resources < min-parallelism-increase && timeSinceLastRescale < scalingIntervalMax` schedule a tryRescale (force a rescale if there is indeed a change in the resource graph at that time in case the last TM crashed) after `scalingIntervalMax`

@zentol you were the one who proposed the addition of `scalingIntervalMax` in the FLIP discussion thread. Do you prefer solution 1 or solution 2 ?

zentol replied:
I don't really see a difference between the 2 options. In both options you had a running job, got resources that do not exceed min-parallellism-increase, and trigger a rescale later after scalingIntervalMax/pipelineRuntimeRescaleMin (or however it is called).

It just seems like option 2 applies scalingIntervalMax twice; once as a condition to schedule tryRescaled, and another time by using it as the delay.

But I assume that that's a mistake in the description, and the difference is meant to be _when_ you start the timeout.
Option 1 does it when a resource is acquired, the 2nd option after a rescale.

So, for a concrete example: We have a job running at p=1 for 24h straight, have a min increase of 1, and a timeout of 1h.
Now we get another slot at the 24h mark.

With option 1 we rescale after an hour.
With option 2 we rescale right away because the timeout already elapsed within the previous 24 hours.

Option 2 is pretty clever, but I think it might be side-stepping the intentions behind the min increase option and timeout.

The min increase option was added to prevent frequent rescalings when resources trickle in slowly.
A slot arrives, you rescale. Another one arrives 5 minutes later, you rescale again, and so on.
The timeout was meant for the case where no further resources arrive; you're holding on to this 1 slot _forever_, but using it would pay off over time.

If you rescale immediately, you will end up rescaling twice in the case where multiples slots do arrive slowly.

With that in mind I'd still prefer option 1.

1996fanrui replied:
Hi @zentol , thanks for your feedback here.

Maybe you misunderstand here. I just clarify what the difference between @echauchot  and me. The current uncertainty is which cases we want to solve rather than the option or how the code is implemented. Here is a case: 

- Assuming the scalingIntervalMax is 60 miuntes, the `jobmanager.adaptive-scheduler.min-parallelism-increase` is 2, and the expected paralleslim is 100.
- The job starts at 09:00:00, and it run with parallelism 99.
- At 09:40:00, the last TM comes.

For this case, the runningTime is 40 minutes, it's less than scalingIntervalMax(60 miuntes). And the resource diff is 1 , it's less than 2 (`jobmanager.adaptive-scheduler.min-parallelism-increase` ).

- @echauchot prefer ignore this case, and never rescale.
- I prefer cover this case, the solution is schedule a rescale after `scalingIntervalMax` or `scalingIntervalMax - runningTime`.

The solution can be ignored now, what do you think about this case? Should we cover this case?

@echauchot Please correct me if I'm wrong, thanks~

echauchot replied:
Thanks Chesnay for your views (again) ! 
> With option 2 we rescale right away because the timeout already elapsed within the previous 24 hours.

This is what I proposed indeed, but you're right that

> it might be side-stepping the intentions behind the min increase option and timeout

if for example after the 24h resources start to arrive 1 slot every 5 min, with this scenario we will restart every 5 min (as the timeout is exceeded) which we want to avoid. So it is better to schedule a timeout when resources arrive and `added ressource < min increase`. When the timeout fires we do one single rescale that takes all the added slots in one shot.
I'll do that

zentol replied:
> Should we cover this case?

We should absolutely cover that case. The whole `scalingIntervalMax` idea was meant for that exact scenario; you have free slots below the min-parallelism-increase threshold for a long time. If you don't cover that case you just have the cooldown between rescale operations.

zentol replied:
>  with this scenario we will restart every 5 min (as the timeout is exceeded min-increase check is overridden) 

This depends on what your starting point for the timeout is. It shouldn't be fixed to the start of the job, but be reset after a job was rescaled.
So, slot arrives at 24h mark, job is rescaled, timeout is reset; for the next hour we only scale up if min-parallel-increase is satisfied, if a slot arrives after that we'd immediately rescale.
I still think this'd be problematic though as described above.


I think a good way to think about it is like this: We probably all had some situation in our live where we thought "oh If only I had done that <some time> earlier." Can be anything, really, let's say it's washing the dishes. You can still do that _now_, but had you done it earlier you may _now_ have been able to use your time better.
That's ultimately what we're doing here.
We look back, see that we had this slot the whole time, and go "ah, if only I had rescaled sooner, but better late than never".
But on the flip-side, if you lost that slot in the mean time and look back you'd think "good job me, that was a good call to ignore that slot!".

This is all about stability; "I could have rescaled and it wouldn't have been a problem.". But you can only conclude that after having out that slot for some period of time; otherwise you're just guessing.

echauchot replied:
> This is all about stability; "I could have rescaled and it wouldn't have been a problem.". But you can only conclude that after having out that slot for some period of time; otherwise you're just guessing.

Exactly. So, when slot arrives scheduling a rescale-check after a timeout seems a good trade-off to ensure that the slot is still there after a period of time. When the timeout fires:
- If the slot is still here we should force-rescale (better late than never case)
- If the slot has vanished we should determine that the target parallelism is the same as current and not rescale (avoid unnecessary rescale case)

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -47,6 +51,10 @@
 class Executing extends StateWithExecutionGraph implements ResourceListener {
 
     private final Context context;
+    private Instant lastRescale = Instant.EPOCH;
```
Actually, I don't understand the `lastRescale` of your implematation. The `Executing` is a new object after each rescale, so `lastRescale` isn't used across multiple rescale.

We should record the `Instant.now()` as the lastRescale at the constructor, right?

echauchot replied:
I agree, there is indeed a bug in Executing being a new object with each rescale when the transition to Restaring and then to Executing state is done. I thought the state of the Executing object was kept between the restarts (with checkpointing) but it is not, a new one is created in the transitionToState(). This leeds to lastRescale being always be EPOCH. Thanks for pointing out.

echauchot replied:
recording lastRescale as Instant.now() in the Executing constructor could work but it means that even a regular transition from Created to Executing state (see [FLIP-160](https://cwiki.apache.org/confluence/display/FLINK/FLIP-160%3A+Adaptive+Scheduler)) would be considered a rescale.

echauchot replied:
Fixed

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -67,13 +77,33 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
         deploy();
 
         // check if new resources have come available in the meantime
         context.runIfState(this, this::maybeRescale, Duration.ZERO);
```
After the `Executing` is created, it means the it's a new rescale. We can rescale after `scalingIntervalMin` even if the new resources have come, right? If yes, we should call `rescaleWhenCooldownPeriodIsOver()` here.

echauchot replied:
not necessary a new rescale it could be a transition from Created to Executing state (FLIP-160). In that case we want to eagerly adapt to resources that could have arrived when while we were starting and not wait for the cooldown period to end.

1996fanrui replied:
Sorry, I don't understand why we should call `maybeRescale` instead of  `rescaleWhenCooldownPeriodIsOver`.

As I mentioned before: `After the Executing is created, it means the it's a new rescale. We can rescale after scalingIntervalMin even if the new resources have come, right?`

If no, the scalingIntervalMin cannot be respected.

Please correct me if I'm wrong.

1996fanrui replied:
Any progress?

echauchot replied:
when the Executing is initially created (without any restart) meaning when the job transitions from Creating to Executing, what is the point in waiting the cooldown period ? In the case of a restart, the scalingIntervalMin is respected because the triggering of the restart (maybeRescale) was not called before scalingIntervalMin during the life of the pipeline. Then there is no point in waiting again when we recreate the Executing object (we have already waited before the restart).

1996fanrui replied:
When the `Executing` is created, it means this is a new restart, we still need to respect the `scalingIntervalMin`.

If we call `maybeRescale` here, AdaptiveScheduler may not guarantee scalingIntervalMin. Following is an example:

- 2 slots comes when job transitions from Creating to Executing.
- The job will go rescale directly if `maybeRescale` is called, because 2 > `min-parallelism-increase`.
- So this round, the Executing ignore the `scalingIntervalMin`.
- This can happen all the time when 2 new slots arrive and the job transitions from Creating to Executing.

IIUC, when the `Executing` is created, it means this is a new restart, we should ensure it can run at least `scalingIntervalMin`.

Please let me know if my understading is wrong or it's not clear.

echauchot replied:
done

1996fanrui commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -488,6 +488,23 @@ public enum SchedulerType {
+                    .defaultValue(Duration.ofSeconds(30))
+                    // rescaling and let the user increase the value for high workloads
+                    .withDescription(
+                            "Determines the minimum time (in seconds) between scaling operations in reactive mode.");
```
1. `in seconds` can be removed due to the type is duration instead of int. It can be various timeUnit.
2. Is the `in reactive mode` exact? As I understand, this option is related to adaptive scheduler event is the reactive mode is disable.

echauchot replied:
1. if the parameter value is passed by CLI as a String, will the duration be parsed correctly with the correct timeUnit?
2. yes but new resources addition call that triggers the rescale (and the cooldown period) only happens in reactive mode.

echauchot replied:
done

1996fanrui replied:
>> Is the in reactive mode exact? As I understand, this option is related to adaptive scheduler even if the reactive mode is disable.
> 
>  2. yes but new resources addition call that triggers the rescale (and the cooldown period) only happens in reactive mode.

The rescale can happen when adaptive scheduler is enabled with reactive mode[1]. And starting from `flink-1.18`, the rescale can happen when adaptive scheduler is enabled with the rest api of FLIP-291 [1], right? 

[1] https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/deployment/elastic_scaling/#reactive-mode
[2] https://cwiki.apache.org/confluence/display/FLINK/FLIP-291%3A+Externalized+Declarative+Resource+Management

echauchot replied:
doc states "You can also use Adaptive Scheduler without Reactive Mode," fair enough adaptive scheduler and reactive mode can be decoupled and Executing#onNewResourceRequirements belongs to Adaptive Scheduler, but min and max parameters refer to rescaling which happens only when reactive mode is activated, no ?

1996fanrui replied:
I guess the min and max parameters take effect when adaptive scheduler is enabled with the rest api of FLIP-291.

Because I don't see any logic that they're just enabled when reactive mode is activated. And these code is placed at adaptive scheduler, so I guess they take effect once adaptive scheduler is enabled. 

Or you can try it locally, enable adaptive scheduler without reactive mode, and rescale it with the rescale button of FLIP-291 in Flink WebUI.

1996fanrui replied:
No progress? I see the description is still `in reactive mode`.

echauchot replied:
fair enough, done

1996fanrui reviewed (changes requested):
Hi @echauchot , thanks for your update.

I left a couple of comments, please take a look in your free time. Please correct me if anything is wrong, thanks! 

BTW, it's better to find another Committer who is familiar with Adaptive Scheduler to review this PR together. It will be more reliable for stability.

echauchot commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -496,6 +556,8 @@ private final class ExecutingStateBuilder {
                 TestingDefaultExecutionGraphBuilder.newBuilder()
                         .build(EXECUTOR_RESOURCE.getExecutor());
         private OperatorCoordinatorHandler operatorCoordinatorHandler;
+        private Duration scalingIntervalMin = Duration.ZERO;
```
no, this builder is used also for the other tests. So it is initialized with a conservative default (cooldown period disabled). It the cooldown tests, this value is set to > 0

echauchot commented:
@dmvk as you authored part on the code in that part, can you review the PR as well ?

echauchot commented:
> @dmvk as you authored part on the code in that part, can you review the PR as well ?

The discussions went around `scalingIntervalMax`, @zentol was the one who proposed the addition of this parameter in the FLIP discussion thread. So maybe it is not needed that we add both of you as reviewers.

echauchot commented:
> ## CI report:
> * [6fa4048](https://github.com/apache/flink/commit/6fa40484d3725b7b4e4b095171a20f57b72ce7d5) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=53323)
> 
> Bot commands

Can't get why spotless in failing on `JobManagerOptions.java`, I rebased on master and ran spotless:apply locally, no failures.
While I figure out what happens, @1996fanrui @zentol can you take a look, I addressed all the comments (including latest ones on interval-max) ?

echauchot commented:
> > ## CI report:
> > 
> > * [6fa4048](https://github.com/apache/flink/commit/6fa40484d3725b7b4e4b095171a20f57b72ce7d5) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=53323)
> > 
> > Bot commands
> 
> Can't get why spotless in failing on `JobManagerOptions.java`, I rebased on master and ran spotless:apply locally on flink-core, no failures. While I figure out what happens, @1996fanrui @zentol can you take a look, I addressed all the comments (including latest ones on interval-max) ?

Wait, the force rescale after max-interval addition generates a regression on AdaptiveScheduler tests. I'll fix them before next round of review

echauchot commented:
> > > ## CI report:
> > > 
> > > * [6fa4048](https://github.com/apache/flink/commit/6fa40484d3725b7b4e4b095171a20f57b72ce7d5) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=53323)
> > > 
> > > Bot commands
> > 
> > 
> > Can't get why spotless in failing on `JobManagerOptions.java`, I rebased on master and ran spotless:apply locally on flink-core, no failures. While I figure out what happens, @1996fanrui @zentol can you take a look, I addressed all the comments (including latest ones on interval-max) ?
> 
> Wait, the force rescale after max-interval addition generates a regression on AdaptiveScheduler tests. I'll fix them before next round of review

Fixed. @zentol  @1996fanrui PTAL

echauchot commented:
> > > > ## CI report:
> > > > 
> > > > * [6fa4048](https://github.com/apache/flink/commit/6fa40484d3725b7b4e4b095171a20f57b72ce7d5) Azure: [FAILURE](https://dev.azure.com/apache-flink/98463496-1af2-4620-8eab-a2ecc1a2e6fe/_build/results?buildId=53323)
> > > > 
> > > > Bot commands
> > > 
> > > 
> > > Can't get why spotless in failing on `JobManagerOptions.java`, I rebased on master and ran spotless:apply locally on flink-core, no failures. While I figure out what happens, @1996fanrui @zentol can you take a look, I addressed all the comments (including latest ones on interval-max) ?
> > 
> > 
> > Wait, the force rescale after max-interval addition generates a regression on AdaptiveScheduler tests. I'll fix them before next round of review
> 
> Fixed. @zentol @1996fanrui PTAL

Do you have any more review comments guys on the latest changes ?

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/scalingpolicy/EnforceParallelismChangeRescalingController.java:
```diff
@@ -0,0 +1,41 @@
+import org.apache.flink.runtime.scheduler.adaptive.allocator.VertexParallelism;
+
+/**
+ * Simple scaling policy for a reactive mode. It just checks that the new parallelism is different
```
Same question about this `reactive mode`.

https://github.com/apache/flink/pull/22985#discussion_r1320535556

echauchot replied:
please see above comment

1996fanrui replied:
Sorry, I didn't see any new comment related to `reactive mode`, please correct my if I missed any comment, thanks~

https://github.com/apache/flink/pull/22985#discussion_r1320535556

echauchot replied:
in that case, we need to remove `for a reactive mode` in both the scaling policies (`EnforceMinimalIncreaseRescalingController` and `EnforceParallelismChangeRescalingController`)

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/scalingpolicy/EnforceParallelismChangeRescalingController.java:
```diff
@@ -0,0 +1,41 @@
+ * Simple scaling policy for a reactive mode. It just checks that the new parallelism is different
+ * (either increase or decrease) from the current parallelism.
+ */
+public class EnforceParallelismChangeRescalingController implements RescalingController {
```
Please add some tests for this class, thanks~

echauchot replied:
ok

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -67,13 +77,36 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
                 executionGraph.getState() == JobStatus.RUNNING, "Assuming running execution graph");
+        this.scalingIntervalMin = scalingIntervalMin;
+        this.scalingIntervalMax = scalingIntervalMax;
+        this.lastRescale =
+                Instant.now(); // Executing is recreated with each restart (when we rescale)
+        // we consider the first execution of the pipeline as a rescale event
```
nit: Would you mind moving the comment to the previous line?

echauchot replied:
done

1996fanrui reviewed (changes requested):
Thanks @echauchot  for the update, I left some comments, please take a look in your free time, thanks~

echauchot commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -488,6 +488,23 @@ public enum SchedulerType {
                                             code(SchedulerExecutionMode.REACTIVE.name()))
                                     .build());
 
+    @Documentation.Section(Documentation.Sections.EXPERT_SCHEDULING)
```
Agree considering that this parameter also deals with job restart.

echauchot replied:
done

echauchot commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -488,6 +488,23 @@ public enum SchedulerType {
+                    .withDescription(
+                            "Determines the minimum time between scaling operations in reactive mode.");
+
+    @Documentation.Section(Documentation.Sections.EXPERT_SCHEDULING)
```
done

echauchot commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -488,6 +488,23 @@ public enum SchedulerType {
+                    .durationType()
+                    .noDefaultValue()
+                    .withDescription(
+                            "Allow the user to configure the time after which a scaling operation is triggered regardless if the requirements are met");
```
I did minor rephrasing but fair enough.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -124,23 +157,74 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
 
+    /**
+     * Rescale the job if added resource meets {@link JobManagerOptions#MIN_PARALLELISM_INCREASE}.
+     * Otherwise, force a rescale after {@link JobManagerOptions#SCHEDULER_SCALING_INTERVAL_MAX} if
+     * the resource is still there.
+     */
     private void maybeRescale() {
-        if (context.shouldRescale(getExecutionGraph())) {
-            getLogger().info("Can change the parallelism of job. Restarting job.");
+        rescaleScheduled = false;
+        if (context.shouldRescale(
+                getExecutionGraph(), false)) { // JobManagerOptions#MIN_PARALLELISM_INCREASE met
```
`JobManagerOptions#MIN_PARALLELISM_INCREASE` is the implementation inside of  `context.shouldRescale`, and we should interface-oriented rather than implementation-oriented, 

So here shouldn't mention `MIN_PARALLELISM_INCREASE`, we just care about `forceRescale` is ture or false here. Does it make sense?

echauchot replied:
Totally agree, thanks for pointing out

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -124,23 +157,74 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
                     Duration.ofMillis(0L),
                     getFailures());
+        } else if (scalingIntervalMax
+                != null) { // JobManagerOptions#MIN_PARALLELISM_INCREASE not met
```
I see all comments and code aren't same line in flink repo. Please move these comments before the code, and update the comments using the `forceRescale`.

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -67,13 +77,36 @@ class Executing extends StateWithExecutionGraph implements ResourceListener {
     }
 
+    @VisibleForTesting
+    void setLastRescale(Instant lastRescale) {
```
Can the `lastRescale` be passed from constructor? 

If so, the `lastRescale` can be final, it's safer. I see it can works, please help check, thanks~

echauchot replied:
No because `lastRescale` is never passed from the outside of the Executing object in production code: it is only initialized at the creation of Executing. `setLastRescale` is only used in test code. Changing a production constructor just for a test case is a bad design.

1996fanrui replied:
> Changing a production constructor just for a test case is a bad design.

So, do you think change a filed from `final` to `non-final` is a good design? If the lastRescale is passed from constructor, the `lastRescale` can be final, it's safe and reasonable. The caller of `new Executing()` can call the  `Instant.now()`.

It is worth mentioning that this suggestion is not created by me, but is a standard in the Java industry. It is called by everyone: `Inject dependencies`. `Inject dependencies` is also mentioned in Flink's "Code Style and Quality Guide", which is in the chapter on `Design for Testability`[1]. Two of them describe the current code:

- Reusability becomes easier if constructors don’t create their dependencies (the objects assigned to the fields), but accept them as parameters.
- Effectively, constructors should have no new keyword.

So, I prefer the `lastRescale` is passed the the `Executing` constructor instead of new it in the `Executing` constructor.

[1] https://flink.apache.org/how-to-contribute/code-style-and-quality-common/#design-for-testability

echauchot replied:
Yes I know this and I agree but the big difference to me is that this field is a **constant** in the production code. But fair enough, I'll do the change if you want to.

echauchot replied:
done

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -570,8 +659,9 @@ public void setHowToHandleFailure(Function<Throwable, FailureResult> function) {
-        public void setCanScaleUp(Supplier<Boolean> supplier) {
-            this.canScaleUp = supplier;
+        public void setCanScaleUp(Boolean minIncreaseMet, Boolean parallelismChangeAfterTimeout) {
+            this.minIncreaseMet = minIncreaseMet;
+            this.parallelismChangeAfterTimeout = parallelismChangeAfterTimeout;
```
Same comment with prod comment[1], Executing side just care about whether `forceRescale` is ture or false, and shouldn't mention minIncreaseMet or parallelismChangeAfterTimeout.

That's why the old code is `canScaleUp` instead of `minIncreaseMet`. I prefer rename them to `canScaleUpWithoutForce` and `canScaleUpWithForce`.

Also, these 2 options can be set by 2 setter methods. 

WDYT?

[1] https://github.com/apache/flink/pull/22985#discussion_r1356382291

echauchot replied:
Agree for the renaming. But regarding the setters I disagree because the 2 parameters are always set together: it is their combination that gives the output of the test. I find it way clearer to set them together.

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -252,28 +257,96 @@ public void testTransitionToFinishedOnSuspend() throws Exception {
+    }
+
+    @Test
+    public void testNotifyNewResourcesAvailableWithMinMetTransitionsToRestarting()
```
And I see a lot of method name and comments include `MinMet`, `MinNotMet`, `min met` and `min not met`, etc.

Would you mind updating all of them to `withoutForce` or `withForce`?

echauchot replied:
done. I agree that keeping high level semantics rather than implementation semantics is good but I feel like with this renaming we loose the real use case semantics that these tests are assessing: resources too small, resources lost etc...

echauchot commented:
@1996fanrui thanks for your comments, please ping me when you're done with all the comments so that I address them all at once and push the code.

1996fanrui commented:
> @1996fanrui thanks for your comments, please ping me when you're done with all the comments so that I address them all at once and push the code.

It's done for now, I will check whether the unit test is enough after your update.

echauchot commented:
@1996fanrui thanks for your review, I'm done addressing all your comments. I hope it's the last round of review.

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/scalingpolicy/EnforceParallelismChangeRescalingControllerTest.java:
```diff
@@ -0,0 +1,64 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Tests for the {@link RescalingController}. */
+@ExtendWith(TestLoggerExtension.class)
```
```suggestion
```



It's not needed, the `TestLoggerExtension` has been added for all test classes by default.

echauchot replied:
done, thanks

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/scalingpolicy/EnforceParallelismChangeRescalingControllerTest.java:
```diff
@@ -0,0 +1,64 @@
+
+/** Tests for the {@link RescalingController}. */
+@ExtendWith(TestLoggerExtension.class)
+public class EnforceParallelismChangeRescalingControllerTest {
```
```suggestion
class EnforceParallelismChangeRescalingControllerTest {
```

The `public` is not needed for junit5.

echauchot replied:
done, thanks

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -594,8 +683,12 @@ public FailureResult howToHandleFailure(Throwable failure) {
+            if (forceRescale) {
+                return canScaleUpWithForce;
+            } else {
+                return canScaleUpWithoutForce;
```
From `@Nullable private Boolean canScaleUpWithForce = false;`, we can see the `canScaleUpWithForce` can be null.

It will be `NullPointException` once it's null here, because the return value is `boolean` instead of `Boolean`.

I prefer remove the `Boolean`, and still mark the default value is false. WDYT?

echauchot replied:
I was trying to stick to nullable as in the production code and also because in some tests we don't care about forcing value. But yes I can definitely move to boolean instead of Boolean to avoid NPE when unboxing.

echauchot replied:
done

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -124,23 +158,70 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
                     getOperatorCoordinatorHandler(),
                     Duration.ofMillis(0L),
                     getFailures());
+        } else if (scalingIntervalMax != null) {
+            getLogger()
+                    .info(
+                            "The longer the pipeline runs, the more the (small) resource gain is worth the restarting time. "
+                                    + "Last resource added does not meet {}, force a rescale after {} time({}) if the resource is still there.",
+                            JobManagerOptions.MIN_PARALLELISM_INCREASE,
+                            JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX.key(),
+                            scalingIntervalMax);
+            // schedule a force rescale in JobManagerOptions.SCHEDULER_SCALING_INTERVAL_MAX time
+            context.runIfState(this, this::forceRescale, scalingIntervalMax);
+        }
```
I still have a question for `scalingIntervalMax`.

> if minimum scaling requirements are not met but last rescale was done more than scaling-interval.max ago, a rescale is forced.

I copy it from FLIP, from it, we can see if last rescale was done more than scaling-interval.max ago, we should call forceRescale, right?

For current code, we didn't have this logic. We always schedule it after scalingIntervalMax. If the `timeSinceLastRescale()` > `scalingIntervalMax`, the delay isn't necessary, right?

echauchot replied:
As you mentioned, we discussed a lot the `scalingIntervalMax` semantics and these semantics have changed during these discussions. I have not updated the FLIP to reflect these changes.
That being said, I agree I if `timeSinceLastRescale() > scalingIntervalMax` we can run an immediate force rescale. I'll update the FLIP as well

echauchot replied:
done

1996fanrui reviewed (changes requested):
> I'm done addressing all your comments. 

Sorry, I see that some comments have not been addressed yet.

> I hope it's the last round of review.

I also very much hope that this PR can be merged as soon as possible. But:

- we did spend a lot of time discussing `scalingIntervalMax` in the early stages of code 
review.
- Judging from your reply, we still have no consensus on test or code style.
- I still have a question for `scalingIntervalMax`, my understanding is that the documentation and code don't seem to be consistent.

echauchot reviewed (commented):
@1996fanrui thanks for the review. Everything is addressed, I think we are in sync, the only thing left is that I don't think we should call rescaleWhenCooldownOver in the Executing creator. See my comment

1996fanrui commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -252,28 +259,94 @@ public void testTransitionToFinishedOnSuspend() throws Exception {
+
+    @Test
+    public void testNotifyNewResourcesAvailableWithCantScaleUpWithoutForceAndCanScaleUpWithForce()
+            throws Exception {
```
- if minimum scaling requirements are not met
  - case1: if last rescale was done more than scaling-interval.max ago, a rescale is forced.
  - case2: otherwise, schedule a forced rescale in scaling-interval.max

This test is case1, and the case2 isn't covered, right?

echauchot replied:
yesterday we added an immediate force rescale if `timeSinceLastRescale > interval-max`. This case is not covered by the tests. The current test covers the previous behaviour in which we scheduled a force rescale no matter `timeSinceLastRescale`. I added a test for immediate force

1996fanrui reviewed (changes requested):
Thanks @echauchot for the update!

The CI failure isn't caused by this PR, please rebase the master branch.

echauchot commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:
```diff
@@ -1162,16 +1167,24 @@ private ExecutionGraph createExecutionGraphAndRestoreState(
+     */
     @Override
-    public boolean shouldRescale(ExecutionGraph executionGraph) {
+    public boolean shouldRescale(ExecutionGraph executionGraph, boolean forceRescale) {
```
sure, thanks for pointing out !

echauchot replied:
is this really needed? The only thing that this boolean parameter adds is a change of `RescallingController`. Both the RescallingControllers are covered by tests and the behavior of `forceRescale` is tested in the `ExecutingTest` class.

1996fanrui replied:
> The only thing that this boolean parameter adds is a change of RescallingController. 

That's exactly I want to test. Check whether the right RescallingController is selected when `forceRescale` is true and false. IIUC, it's tested for now.

> Both the RescallingControllers are covered by tests and the behavior of forceRescale is tested in the ExecutingTest class.

IIUC, you have tested 2 types of RescallingControllers individually. In ExecutingTest, you just mock the `shouldRescale()` and the real `shouldRescale` isn't tested.

Please correct me if my understanding is wrong, thanks~

1996fanrui replied:
I can accept it if you still think it's not needed. Up to you, thanks

echauchot replied:
> That's exactly I want to test. Check whether the right RescallingController is selected when forceRescale is true and false. IIUC, it's tested for now.

well, testing that is equivalent to testing the ternary expression ```
RescalingController rescalingControllerToUse =
                                    forceRescale ? forceRescalingController : rescalingController;
``` which is a JDK built-in. As a consequence, that still seems unnecessary to me.

echauchot commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -252,28 +259,94 @@ public void testTransitionToFinishedOnSuspend() throws Exception {
+                            .setScalingIntervalMax(Duration.ofSeconds(1L))
+                            .build(ctx);
+            ctx.setCanScaleUp(
+                    false, false); // => schedule force rescale but resource lost on timeout =>
```
done

echauchot commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -252,28 +259,94 @@ public void testTransitionToFinishedOnSuspend() throws Exception {
+            ctx.setCanScaleUp(true); // => rescale
+            ctx.setExpectRestarting( // scheduled rescale should restart the job after cooldown
+                    restartingArguments -> {
+                        assertThat(restartingArguments.getBackoffTime(), is(Duration.ZERO));
```
As said in the PR and in a commit comment, this is done on purpose: MockExecutingContext#runIfState does not do delays but runs the task immediately and sets a boolean to check that an async task was scheduled. This is to avoid having long tests.

1996fanrui replied:
Thanks for your clarification!

echauchot commented:
> Thanks @echauchot for the update!
> 
> The CI failure isn't caused by this PR, please rebase the master branch.
> 
> Also, would you mind squash all commits to one? They are too much, and I believe this PR is close to being merged.

I have rebased and pushed. I'll squash when the review is finished.

1996fanrui reviewed (commented):
Hi @echauchot , I have only one question[1] for this pr, pleade help take a look in your free time, thanks for your patience, discussion and updating!

Hope this pr can be merged within one week!

[1] https://github.com/apache/flink/pull/22985#discussion_r1320535241

echauchot reviewed (commented):
All done and squashed to one commit. Thanks

1996fanrui reviewed (approved):
Thanks @echauchot  for the hard work!

LGTM, I will merge it if no any comments within 2 days.

1996fanrui commented:
Would you mind updating the commit message to `[FLINK-21883][scheduler] Implement cooldown period for adaptive scheduler`?

The current commit message missed the `module name` and `adaptive scheduler`.

echauchot commented:
> Thanks @echauchot for the hard work!
> 
> LGTM, I will merge it if no any comments within 2 days.

Thanks for your review !
